File: android/src/main/kotlin/aayushsharma/me/flutter_comscore/FlutterComscorePlugin.kt
package aayushsharma.me.flutter_comscore

import android.content.Context
import android.content.pm.PackageManager
import android.util.Log
import androidx.annotation.NonNull
import com.comscore.*
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement

/** FlutterComscorePlugin */
class FlutterComscorePlugin: FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel
  lateinit var applicationContext: Context
  private val PUBLISHER_ID_KEY = "com.flutterComscore.PUBLISHER_ID"
  private val LOGNAME = "FlutterComscorePlugin"
  lateinit var publisherId: String

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    applicationContext = flutterPluginBinding.applicationContext
    getPublisherId(applicationContext)
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_comscore")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    val args = call.arguments as String?
    when(call.method) {
      "setup" -> {
        Log.e(LOGNAME, "starting comscore: $args")
        val startComscoreArgs = Json.decodeFromString<StartComscoreArgs>(args?:"")

        startComscore(applicationContext, startComscoreArgs)

        result.success(null)
      }
      "notifyViewEvent" -> {
        try {
          val data = Json.decodeFromString<NotifyViewEventArgs>(args?:"")
          notifyViewEvent(data.category, data.eventData)
        } catch(e: Exception) {
          Log.e(LOGNAME, "notifyViewEvent error: $e")
          result.error(LOGNAME, e.toString(), null)
        }
      }
      "setUserConsent" -> {
        try {
          setUserConsent(args?.toInt() ?: 0)
        } catch (e: Exception) {
          Log.e(LOGNAME, e.toString())
          result.error(LOGNAME, e.toString(), null)
        }
      }
      "notifyBackgroundUXStart" -> notifyUxActive()
      "notifyBackgroundUXStop" -> notifyUxInactive()
      else -> result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  private fun getPublisherId(context: Context) {
    try {
      val applicationInfo = context.packageManager.getApplicationInfo(context.packageName, PackageManager.GET_META_DATA)
      val metaData = applicationInfo.metaData
      publisherId = metaData.getInt(PUBLISHER_ID_KEY)!!.toString()

    } catch(e: Exception) {
      Log.e(LOGNAME, "Error fetching publisher id: $e")
    }
  }

  private fun startComscore(context: Context, startComscoreArgs: StartComscoreArgs) {
    // Get the publisher id from manifest
    try {
      var labels = HashMap<String,String>()
      labels["cs_ucfr"] = startComscoreArgs.userConsent?.toString() ?: ""

      val publisherConfig = PublisherConfiguration.Builder()
        .publisherId(publisherId)
        .persistentLabels(labels)
        .build()
      Analytics.getConfiguration().addClient(publisherConfig)

      Analytics.getConfiguration().setUsagePropertiesAutoUpdateMode(UsagePropertiesAutoUpdateMode.FOREGROUND_AND_BACKGROUND)

      if(startComscoreArgs.isChildDirected) {
        Analytics.getConfiguration().enableChildDirectedApplicationMode()
      }

      if(startComscoreArgs.debug) {
        Analytics.getConfiguration().enableImplementationValidationMode()
      }

      Analytics.start(context)
    } catch (e: Exception) {
      Log.e(LOGNAME, "Error starting comscore: $e")
    }
  }

  private fun notifyUxActive() {
    try {
      Analytics.notifyUxActive()
    } catch (e: Exception) {
      Log.e(LOGNAME, "Error notifyUxActive comscore: $e")
    }
  }

  private fun notifyUxInactive() {
    try {
      Analytics.notifyUxInactive()
    } catch (e: Exception) {
      Log.e(LOGNAME, "Error notifyUxInactive comscore: $e")
    }
  }

  private fun notifyViewEvent(category: String,eventData: HashMap<String,String>?) {
    try {
      var labels = HashMap<String,String>()
      labels.putAll(eventData?: emptyMap())
      labels["ns_category"] = category
      Analytics.notifyViewEvent(labels)
    } catch (e: Exception) {
      Log.e(LOGNAME, "Error sending notifyview event: $e")
    }
  }

  private fun setUserConsent(userConsent: Int) {
    // Get the publisher id from manifest
    try {
      Analytics.getConfiguration().getPublisherConfiguration(publisherId).setPersistentLabel("cs_ucfr", userConsent.toString())
      Analytics.notifyHiddenEvent()
    } catch (e: Exception) {
      Log.e(LOGNAME, "Error starting comscore: $e")
    }
  }
}

File: ios/Classes/FlutterComscorePlugin.swift
import Flutter
import UIKit
import ComScore

public class FlutterComscorePlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "flutter_comscore", binaryMessenger: registrar.messenger())
        let instance = FlutterComscorePlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "setup":
            guard let args = call.arguments as? String else {
                result("Empty arguments")
                return
            }
            
            do {
                try startComscore(args: JSONDecoder().decode(StartComscoreArgs.self, from: Data(args.utf8)))
            } catch {
                result("Error starting comscore: \(error)")
                return
            }
            
            result(nil)
        case "notifyViewEvent":
            guard let args = call.arguments as? String else {
                result("Empty arguments")
                return
            }
            
            do {
                let notifyViewEventArgs = try JSONDecoder().decode(NotifyViewEventArgs.self, from: Data(args.utf8))
                notifyViewEvent(category: notifyViewEventArgs.category, eventData: notifyViewEventArgs.eventData)
            } catch {
                result("Error comscore notifyingview event: \(error)")
                return
            }
            
            result(nil)
        case "notifyBackgroundUXStart":
            notifyBackgroundUXStart()
            result(nil)
        case "notifyBackgroundUXStop":
            notifyBackgroundUXStop()
            result(nil)
        case "setUserConsent":
            guard let args = call.arguments as? String else {
                result("Empty arguments")
                return
            }
            
            do {
                try setUserConsent(userConsent: args)
            } catch {
                result("Error comscore notifyingview event: \(error)")
                return
            }
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func startComscore(args: StartComscoreArgs) throws {
        let infoDict = Bundle.main.infoDictionary
        guard let infoDict, let pubId = infoDict["FLUTTER_COMSCORE_PUBLISHER_ID"] as? String else {
            throw ComscoreError.pubIdNotFound
        }
        
        let myPublisherConfig = SCORPublisherConfiguration(builderBlock: { builder in
            builder?.publisherId = pubId
            
            if let userConsent = args.userConsent {
                builder?.persistentLabels = ["cs_ucfr": userConsent]
            }
        })
        
        SCORAnalytics.configuration().addClient(with:myPublisherConfig)
        SCORAnalytics.configuration().usagePropertiesAutoUpdateMode = .foregroundAndBackground
        
        if args.isChildDirected {
            SCORAnalytics.configuration().enableChildDirectedApplicationMode()
        }
        
        if args.debug {
            SCORAnalytics.configuration().enableImplementationValidationMode()
        }
        
        SCORAnalytics.start()
    }
    
    private func notifyViewEvent(category: String, eventData: [String: String]?) {
        var labels: [String: String] = [:]
        labels["ns_category"] = category
        if let eventData {
            labels.merge(eventData) { (current, new) in new }
        }
        
        SCORAnalytics.notifyViewEvent(withLabels: labels)
    }
    
    private func notifyBackgroundUXStart() {
        SCORAnalytics.notifyUxActive()
    }
    
    private func notifyBackgroundUXStop() {
        SCORAnalytics.notifyUxInactive()
    }
    
    private func setUserConsent(userConsent: String) throws {
        let infoDict = Bundle.main.infoDictionary
        guard let infoDict, let pubId = infoDict["FLUTTER_COMSCORE_PUBLISHER_ID"] as? String else {
            throw ComscoreError.pubIdNotFound
        }
        
        SCORAnalytics.configuration().publisherConfiguration(withPublisherId: pubId).setPersistentLabelWithName("cs_ucfr", value: userConsent)
    }
    
    enum ComscoreError: Error {
        case pubIdNotFound
    }
}
